LinGui: make QSV encoding actually work

jstebbins opened the pull request:
Assuming you built HandBrake with 'configure --enable-qsv' and you have
built and installed Intel MediaSDK in a directory that is in your LD search
path, the QSV encoders now work.  HW decode is not supported.

**Description of Change:**
A VA display must be initialized and given to libmfx after MFXInit() and before anything else.

**Test on:**

- [x] Windows 10+  (via MinGW)
- [x] Ubuntu Linux

Should be tested on windows since I changed things slightly in the windows code path.

jstebbins committed:
LinGui: make QSV encoding actually work

Assuming you built HandBrake with 'configure --enable-qsv' and you have
built and installed Intel MediaSDK in a directory that is in your LD search
path, the QSV encoders now work.  HW decode is not supported.

sr55 commented:
Interesting. I thought a user had reported this working :/

jstebbins commented:
I've never seen any definitive reports of it working.  But I've not been very on top of it.

sr55 commented:
@maximd33 - Any chance you could take a look?

maximd33 commented:
let me comment more next week, thank you

about reports: like https://github.com/HandBrake/HandBrake/issues/1701#issuecomment-446726757

jstebbins commented:
Interesting. The log from #1701 shows this
```
[08:39:49] gtkgui: HandBrake 1.1.2 (2018111300) - Linux x86_64 - https://handbrake.fr
libva info: VA-API version 1.1.0
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib/dri/i965_drv_video.so
libva info: Found init function __vaDriverInit_1_1
libva info: va_openDriver() returns 0
```
I think something has changed in either libmfx or mediasdk because those libva logs are no longer printed unless I manually create and initialize a VADisplay myself.  Those libva messages are printed when vaInitialize() is called and I don't see vaInitialize getting called in either libmfx or mediasdk in current code.  libmfx has a mfx_va_glue.c file with a mfx_allocate_va() function that would do this, but it is not called anywhere.  mfx_allocate_va() is not terribly user friendly when you have multiple display adapters anyway.  User would have to set an env variable to tell it the right adapter to use.  This PR searches for the right one, no user action required.

sr55 commented:
I did update mfx more recently so maybe changed then?

jstebbins commented:
Another interesting note, when I use the iHD driver, I get a crash when using 'quality' preset.  I don't get a crash when using i965 driver.

When using the medium or speed presets, the call to MfxHwH264Encode::CmContext::Setup where the crash occurs doesn't even happen.  A different code path apparently.

Backtrace:
```
(gdb) bt
#0  0x00007ffff5e7153f in raise () at /lib64/libc.so.6
#1  0x00007ffff5e5b895 in abort () at /lib64/libc.so.6
#2  0x00007ffff5e5b769 in _nl_load_domain.cold.0 () at /lib64/libc.so.6
#3  0x00007ffff5e699f6 in .annobin_assert.c_end () at /lib64/libc.so.6
#4  0x00007fffae5aec67 in MfxHwH264Encode::CmRuntimeError::CmRuntimeError()
    (this=0x7fffa000a760)
    at /home/jstebbins/Source/MediaSDK/_studio/mfx_lib/encode_hw/h264/include/mfx_h264_encode_cm.h:65
#5  0x00007fffae63a0ba in MfxHwH264Encode::CmContext::Setup(MfxHwH264Encode::MfxVideoParam const&, CmDevice*, VideoCORE*) (this=0x7fffa0b911e0, video=..., 
    cmDevice=0x7fffa0128110, core=0x7fffa00fc6f0)
    at /home/jstebbins/Source/MediaSDK/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_cm.cpp:848
#6  0x00007fffae639e6b in MfxHwH264Encode::CmContext::CmContext(MfxHwH264Encode::MfxVideoParam const&, CmDevice*, VideoCORE*)
    (this=0x7fffa0b911e0, video=..., cmDevice=0x7fffa0128110, core=0x7fffa00fc6f0)
    at /home/jstebbins/Source/MediaSDK/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_cm.cpp:831
#7  0x00007fffae5a1049 in MfxHwH264Encode::ImplementationAvc::Init(mfxVideoParam*) (this=0x7fffa014b910, par=0x7fffa00f83b0)
    at /home/jstebbins/Source/MediaSDK/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_hw.cpp:1049
#8  0x00007fffae59d5ce in MFXHWVideoENCODEH264::Init(mfxVideoParam*)
    (this=0x7fffa0128340, par=0x7fffa00f83b0)
    at /home/jstebbins/Source/MediaSDK/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_encode_hw.cpp:188
#9  0x00007fffae39ee2d in MFXVideoENCODE_Init(mfxSession, mfxVideoParam*) (
    session=0x7fffa00fc518, par=0x7fffa00f83b0)
    at /home/jstebbins/Source/MediaSDK/_studio/mfx_lib/shared/src/libmfxsw_encode.cpp:707
#10 0x00000000027f93ff in MFXVideoENCODE_Init(mfxSession, mfxVideoParam*)
    (session=0x7fffa00fc518, par=0x7fffa00f83b0)
    at src/mfx_exposed_functions_list.h:62
#11 0x00000000004bc600 in encqsvInit (w=0x7fffa000a460, job=0x7fffa0031e00)
    at ../libhb/enc_qsv.c:1118
#12 0x00000000004a7dda in do_job (job=0x7fffa0031e00) at ../libhb/work.c:1758
#13 0x00000000004a48b1 in work_func (_work=0x5241240) at ../libhb/work.c:154
#14 0x000000000046f7d3 in hb_thread_func (_t=0x4816770) at ../libhb/ports.c:868
#15 0x00007ffff73fc58e in start_thread () at /lib64/libpthread.so.0
#16 0x00007ffff5f366a3 in clone () at /lib64/libc.so.6
```

maximd33 commented:
>when I use the iHD driver, I get a crash when using 'quality' preset.

@jstebbins are you using fresh clone from https://github.com/Intel-Media-SDK/MediaSDK or a particular release ?
what HW/OS do you run on?

jstebbins commented:
Using a fresh clone of mediasdk and built git HEAD.  I used tag intel-media-18.3.0 of the media-driver however because that is the most recent version that will build against libva-2.3 (libva provided by fedora).  I didn't want to build and install libva-2.4, but I can if you think that might be contributing to the problem.

OS Fedora 29
CPU is i7-7700HQ, and nvidia GTX 1060 mobile is the primary display adapter.

I'm thinking I need to add to my patch to programatically set LIBVA_DRIVER_NAME=iHD.  What actually happens when it uses the i965 va driver?  Are settings and features being silently ignored when using this driver?

jstebbins commented:
@sr55 I cloned Luca's mfx_dispatch and checked the history, it appears @maximd33 removed the VA initialization code in his last update to sync it with 1.27.  I'm not sure if that was intentional, but this PR handles multi-adapter scenario better that the old mfx_dispatch VA code.  So if that code is to be restored in mfx_dispatch, I'll probably roll these changes into a PR agains mfx_dispatch.

jstebbins commented:
So, aside from the various bugs, the main problem with QSV support on linux is distribution of the required dependencies. I think I have a nice solution to this. Looking at how MediaSDK works, I'm pretty sure that I can package up *everything* that is needed for QSV support on linux into a flatpak plugin for HandBrake.  I would have to dlload libmfx and modify how those functions get called in libhb, but it would be a pretty slick way to distribute all the required external shared libs that QSV relies on.  My flatpak builds already have a plugin interface defined on them (I made my own plugins for dvdcss and makemkv's decryption lib for transparent decryption). This plugin can be maintained completely separately.  Users can install it if they want to try QSV and support will magically appear in HandBrake.

ffmpeg does not dlopen libmfx, so I can't enable mfx in ffmpeg in this scenario. But it appears that this was only for decode support, and at least for now I don't plan to allow hw decode on linux.

jstebbins commented:
The most recent commit here adds a new configure option --enable-qsv-dll that enables dynamic loading of libmfx and runtime detection of it's presence.  Linux enables this feature automatically when --enable-qsv is used.  Windows can optionally enable using --enable-qsv-dll.

This option makes it so Linux can be distributed with QSV enabled, but without all of MediaSDK and it's dependencies being built in.  I plan to package MediaSDK and it's dependencies as a flatpak plugin (already have it built, just need to test it).  When the plugin is installed, libmfx will be dlopen'd and used automatically.  End users can also obtain or build their own copy of MediaSDK an it'll be detected at runtime for non-flatpak installs.

This also allows use of an unmodified version of MediaSDK for linux which *only* builds a shared version of libmfx, no static.

jstebbins commented:
I forgot to mention, there is one nit I haven't covered yet.  If --enable-qsv-dll is enabled for the windows build, hardware decoding must be disabled because ffmpeg doesn't dlopen libmfx. So we must disable libmfx in ffmpeg if we want to dynamic load it.  When --enable-qsv-dll is set, --enable-libmfx is *not* set when configuring ffmpeg.  This may be enough to disable hardware decoding in libhb, but I haven't yet checked to be sure that is true.

jstebbins commented:
@maximd33 the mfx dispatcher that comes with the open source mediasdk https://github.com/Intel-Media-SDK/MediaSDK is linux only, no windows support.  Is there an open source version supplied by Intel somewhere that includes windows support?  Where do windows developers normally go for this?

maximd33 commented:
for Windows we still need to use fork from Luca OR from any Intel's distribution,
github - let me think if can be added under https://github.com/Intel-Media-SDK/MediaSDK/tree/master/api/mfx_dispatch ...

ffmpeg's decode: for now it is fine
but WIP to enable full decode/encode been HW accelerated

jstebbins committed:
qsv: disable xlib in ffmpeg

It is autodetected and results in link failure if ffmpeg finds it
bacause we do not add it to final link flags

jstebbins committed:
flatpak: allow building with QSV enabled

jstebbins committed:
flatpak: add Intel MediaSDK plugin

Defines a flatpak manifest that is used to download and build all the
dependencies for MediaSDK 19.1.pre2.  These are built into a flatpak
plugin that can be installed independently from the main HandBrake flatpak.

When the plugin is installed, HandBrake gains the ability to do QSV
encoding with the latest MediaSDK and VA drivers.

When HandBrake is configured with '--flatpak --enable-qsv' and
'make pkg.create.flatpak' is run, both the HandBrake flatpak and the
MediaSDK plugin will be built.

The name of the MediaSDK plugin file is:
fr.handbrake.plugin.IntelMediaSDK-$(HB.version)-$(HB.machine).flatpak

I've also changed the the name of the HandBrake flatpak file to conform
better with flatpak naming:
fr.handbrake.ghb-$(HB.version)-$(HB.machine).flatpak
fr.handbrake.HandBrakeCLI-$(HB.version)-$(HB.machine).flatpak

jstebbins commented:
After discussing with @maximd33, I've decided dlopen'ing libmfx.so isn't such a good idea at this time. So I backed out the DLL stuff.  Keeping it on a separate branch for now.

I do have a working config for building the flatpak with QSV enabled and building a separate MediaSDK plugin that contains all the necessary dependencies to run the very latest in the flatpak sandbox. Aside from the one problem I still have with h.264 'quality' preset failing to initialize, everything else seems to be working OK.  All the h.265 8 and 10 bit presets work as expected.

So I think I'm happy with the state of this PR now.

bradleysepos commented:
I can't test due to lack of QSV hardware, but wanted to say nice work. 👍

jstebbins commented:
I mentioned it in a commit message, but I should bring it up here as well.  I changed the names of the flatpak files that are created when packaging flatpaks.  This may require a change to the nightly build script.  The names now follow the convention used for flatpak IDs.  e.g. GUI flatpak
```fr.handbrake.ghb-20190224171000-e950fbe-qsv-x86_64.flatpak``` 
and the name of the new mediasdk plugin 
```fr.handbrake.plugin.IntelMediaSDK-20190224171000-e950fbe-qsv-x86_64.flatpak```

jstebbins commented on libhb/qsv_common.c:
```diff
@@ -892,6 +912,7 @@ hb_list_t* hb_qsv_load_plugins(hb_qsv_info_t *info, mfxSession session, mfxVersi
 
     if (HB_CHECK_MFX_VERSION(version, 1, 8))
     {
+#if !defined(SYS_LINUX)
```
Reviewing my own PR :stuck_out_tongue: 
@sr55 and @maximd33 I had to disable this block of code on linux because MFXVideoUSER_Load failed and it seems to work without it. i.e. it finds the h.265 encoder and succeeds at encoding. Do you know why there is a difference between windows and linux here and whether there would be a better way of handling this?

jstebbins replied:
One alternative option would be to make a failure during hb_qsv_load_plugins non-fatal.  Would that work properly on Windows.  I.e. if the error value of hb_qsv_load_plugins were ignored and query_capabilities continued to attempt to call MFXVideoENCODE_Query, would it behave correctly on Windows?

maximd33 replied:
latest Linux/Windows drivers does not require MFXVideoUSER_Load()

making it non-fatal should be fine and should not break backward compatibility
for the legacy driver users, 
over time can be removed

chriscjs commented:
@jstebbins Your qsv branch fails to build on Arch Linux.  I had to symlink /usr/include/drm to /usr/include/libdrm to get it to build.  Otherwise it works for QSV encoding on my Skylake system.  I tested the h265 encoder.  One glitch I noticed was starting off with the quality preset made the encoding fail.  Switching to balanced and it worked.  Then switching back to quality and that worked whereas before it failed.  If I quit and then restart Handbrake I have to go through the same process to get the quality preset to work.

[log.txt](https://github.com/HandBrake/HandBrake/files/2927128/log.txt)

jstebbins commented:
@chriscjs thanks for pointing that out.  /usr/include/drm is in kernel-headers package.  /usr/include/libdrm is in libdrm-devel package (on fedora, arch should have similarly named packages).  drm.h is identical between these 2 packages.  I didn't realize they both existed and just included the first that I saw.  It eliminates a build dependency to use the libdrm version, so I'll switch it to that.

I haven't updated the build dependencies documentation to cover this patch yet.  There are a couple packages to add to the build dependencies when building with --eanble-qsv.
- libdrm-devel
- libva-devel

jstebbins commented:
The preset switching failure is a GUI glitch.  I'll fix that separately since it's not strictly related to qsv and also should be backported to our bugfix branch.

chriscjs commented:
I attached the activity log to my post above.

jstebbins commented:
Anybody want/need more time to review and test?  I think this is pretty much ready.

sr55 commented:
I'll run a quick test on it tomorrow in Windows if you've not done so already?

jstebbins commented:
I haven't.

maximd33 commented:
will test but only at weekend, if you can wait

jstebbins commented:
No hurry.  Just wanted to let everyone know I'm ready.

sr55 commented:
Windows H.264/5 testing done. Seems to be OK.

maximd33 commented:
If s55 is fine - it is already solid confirmation 

> 7 марта 2019 г., в 22:04, Scott <notifications@github.com> написал(а):
> 
> @sr55 approved this pull request.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

jlesage commented:
@jstebbins, thanks for this fix.  The upgrade of libmfx in HandBrake 1.2.1 broke QSV on Linux, so it's great that you restored the feature.

However, with this fix, it is no longer possible to choose the VA-API  user space driver to use.  See https://github.com/HandBrake/HandBrake/issues/1701#issuecomment-448983217.

I think we should force the use of `iHD` only if the `LIBVA_DRIVER_NAME` environment variable is not already set.

jstebbins commented:
Ah, good point.  At the very least it will make testing easier.  I a bit surprised that some older hardware wouldn't be supported by the iHD driver. But it's a valid request regardless.

jlesage commented:
The `iHD` driver support the following CPUs: https://github.com/intel/media-driver#supported-platforms.
So Sandy bridge, Ivy bridge and Haswell CPUs are not supported by this driver.

jstebbins commented:
So, if you are using an older CPU, what is the symptom when loading iHD. Does HandBrake fail to show QSV as an option?  Or does it fall back to software encoding with QSV?  I'm wondering if we can perform a fallback check if iHD fails.

jlesage commented:
Correct, the QSV encoder doesn't show up in the list.

Having a fallback mechanism would be useful. Would you still honour the `LIBVA_DRIVER_NAME` environment variable if set?

If it can help, here are the relevant messages that occur when loading the `iHD` driver fails:

```
libva info: VA-API version 1.4.0
libva info: va_getDriverName() returns 0
libva info: User requested driver 'iHD'
libva info: Trying to open /opt/intel/mediasdk/lib64/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_4
libva error: /opt/intel/mediasdk/lib64/iHD_drv_video.so init failed
libva info: va_openDriver() returns 1
```

jstebbins commented:
> Would you still honour the LIBVA_DRIVER_NAME environment variable if set?

Yes.  IMO, if the user has set the environment variable, we should *only* attempt what they asked for.

Thanks for the log.  That definitely does help.

jlesage commented:
Great, let me know if you want me to test something.

jstebbins commented:
Done  72e71b6

If the env var is set, only the env setting will be attempted.
Else, iHD, then i965, then VA default (no env set) will be attempted.

jlesage commented:
Works well for me:
```
[21:52:14] hb_display_init: attempting VA driver 'iHD'
libva info: VA-API version 1.4.0
libva info: va_getDriverName() returns 0
libva info: User requested driver 'iHD'
libva info: Trying to open /opt/intel/mediasdk/lib64/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_4
libva error: /opt/intel/mediasdk/lib64/iHD_drv_video.so init failed
libva info: va_openDriver() returns 1
[21:52:16] hb_display_init: attempting VA driver 'i965'
libva info: VA-API version 1.4.0
libva info: va_getDriverName() returns 0
libva info: User requested driver 'i965'
libva info: Trying to open /opt/intel/mediasdk/lib64/i965_drv_video.so
libva info: Found init function __vaDriverInit_1_4
libva info: va_openDriver() returns 0
```
Thanks!

jlesage commented:
Btw, were you able to fix the preset switching glitch, as mentioned  in https://github.com/HandBrake/HandBrake/pull/1922#issuecomment-469308120 ?

I also experimented the behaviour described in https://github.com/HandBrake/HandBrake/pull/1922#issuecomment-469302943.  Not sure if it was the right way to fix the issue, by I did the following patch, if it can help:  https://github.com/jlesage/docker-handbrake/blob/master/A00-hb-video-preset.patch

jstebbins commented:
It should already be fixed in git master.  It's been working for me.
See 90c8d60

jlesage commented:
Thank you very much for the info!
